Custom operators and builders (#33355)
idaholab/moose: framework/doc/content/syntax/MFEMProblemComposer/index.md

> # MFEMProblemComposer System

Reviewer: ```suggestion
# ProblemComposer System
```

The `MFEMProblemComposer` is a builder/adapter class that is used for constructing

> problem-operators with custom inputs. This class is specifically intended for composing

Reviewer: ```suggestion
The `ProblemComposer` system allows the user to construct
problem operators with custom inputs. This class is specifically intended for composing
```

user defined custom operators which may be raw `mfem::Operator`'s optimised for specific

> purposes e.g. MHD with customised inputs. The Operators may need thin layer access to the MOOSE

Reviewer: ```suggestion
purposes e.g. MHD with customised inputs. The operators may need thin layer access to the MOOSE
```

multi-physics system. As of yet only a single problem operator per `MFEMProblem` is used, even if

> multiple are defined. The ProblemOperatorBuilder classes are built within the `MFEMProblem` class 

Reviewer: ```suggestion
multi-physics system. As of yet, only a single problem operator object per [MFEMProblem.md] is used (even if
multiple classes are defined). The [problem composer](ProblemComposerBase.md) classes are built within the [MFEMProblem.md] class 
```

> however the `ProblemOperator`s are built and owned by the MFEM executioners.

Reviewer: ```suggestion
however the [ProblemOperator.md]s are built and owned by the MFEM executioners.
```

> ## Steady problem composer

Reviewer 2: ```suggestion
## SteadyProblemComposer
```

The `SteadyProblemComposer` class is the default of the `MFEMSteady` executioner and does
not need to be declared explicitly. The `SteadyProblemComposer` class uses systematic
logic to build one of 3 possible `ProblemOperator`s they are the `EquationSystemProblemOperator`,

> `ComplexEquationSystemProblemOperator` and `EigenproblemESProblemOperator`.

Reviewer: ```suggestion
The [SteadyProblemComposer.md] class is the default for the [MFEMSteady.md] executioner and does
not need to be declared explicitly. The [SteadyProblemComposer.md] class uses systematic
logic to build one of 3 possible `ProblemOperator`s: they are the `EquationSystemProblemOperator`,
`ComplexEquationSystemProblemOperator` and `EigenproblemESProblemOperator`.
```

> ## Transient problem composer

Reviewer 2: ```suggestion
## TransientProblemComposer
```

The `TransientProblemComposer` class is the default of the MFEMTransient executioner and
does not need to be declared explicitly. The `TransientProblemComposer` class builds 

> the `TimeDependentEquationSystemProblemOperator`.

Reviewer: ```suggestion
The [TransientProblemComposer.md] class is the default for the [MFEMTransient.md] executioner and
does not need to be declared explicitly. The [TransientProblemComposer.md] class builds 
the `TimeDependentEquationSystemProblemOperator`.
```

> ## CustomProblemComposer example

Reviewer: ```suggestion
## Using a custom problem composer to plug in a custom problem operator
```

> ## Custom problem operator composer example

Reviewer 2: ```suggestion
## Custom ProblemComposer example
```

> The custom problem operator example will follow the `MFEMCustomProblemOperator.C` unit-test which

Reviewer: This is perhaps the most important comment, and the thought I'll finish with for now: it'd be really, really nice if we could put our custom operator in `test/{src,include}/mfem/problem_operators` and our custom composer in `test/{src,include}/mfem/problem_composers` and then a real input file in `test/tests/mfem/operators` or similar. I also think the custom operator should be as close as it can possibly be to a carbon copy of `ex0p.cpp` maybe with a small extension, e.g. using an auxkernel or a postprocessor, to show potential newcomers how easy it is, how easy it can be for them too, how easy it is to _incrementally_ adopt the functionality we can give them. And then, in this walkthrough we just take excerpts of the scaffolding needed and of the input file from that example. IMHO, we don't need to explain the user how to build an MFEM problem here.

Author: The minimum that would be needed would be to register the grid-functions, not necessarily the FE-spaces in the global MOOSE Object storage, but I think its doable

Author: True we may not do, but I pasted in the example problem so a user who knows how to build one in MFEM knows what it looks like in MOOSE-MFEM without having to jump between the documentation and the example problem

> IMHO, we don't need to explain the user how to build an MFEM problem here

Author: I have added the example test/{include,src}/mfem/problem_{operator,composer}. And a test case has been added to the test harness.

is based on MFEM's [ex0p](https://github.com/mfem/mfem/blob/master/examples/ex0p.cpp). Firstly a
CustomProblemOperator class must be built, if the problem is steady i.e called from the `MFEMSteady` 
executioner then the `ProblemOperator` class inherits from `Moose::MFEM::ProblemOperator`. If the
CustomProblemOperator class is transient i.e. called from the `MFEMTransient` executioner then
the class must inherit from `Moose::MFEM::TimeDependentProblemOperator`. For this case the problem

> is steady and an example class may look as follows:

Reviewer: ```suggestion
This custom problem operator example will follow the `MFEMCustomProblemOperator.C` unit test which
is based on MFEM's [ex0p](https://github.com/mfem/mfem/blob/master/examples/ex0p.cpp). Firstly, a
custom [ProblemOperator.md] class must be built. If the problem is steady, i.e called from the [MFEMSteady.md] 
executioner, then the problem operator class inherits from `Moose::MFEM::ProblemOperator`. If the
problem is transient, i.e. called from the [MFEMTransient.md] executioner, then
the class must inherit from `Moose::MFEM::TimeDependentProblemOperator`. For this case, the problem
is steady and an example class may look as follows:
```

```cpp
class CustomDummyProblemOperator : public Moose::MFEM::ProblemOperator
{
  private:
  .
  .
  .
  public:
    // Constructor
    CustomDummyProblemOperator(MFEMProblem & prob, ...);

    // Solve the equation
    virtual void Solve() override;

    // Mult by the operator

>     void Mult(const mfem::Vector & x, mfem::Vector & y) const override {};

Reviewer: ```suggestion
    // Apply the operator
    void Mult(const mfem::Vector & x, mfem::Vector & y) const override {}
```

};

```

> As the problemOperator is built by the MFEM executioners all registered entities in

Reviewer 2: Please search and correct all remaining instances of camelCase in the documentation that have slipped through (like `problemOperator` here

{
  // Retrieve the FE-space and gridFunction
  const std::string fe_space_name = "h1";
  const std::string grid_function_name = "var0";
  auto fes = prob0.getProblemData().fespaces.GetShared(fe_space_name);
  auto grid_function = prob0.getProblemData().gridfunctions.GetShared(grid_function_name);

  // Boundary conditions
  *grid_function = 0.00;
  fes->GetBoundaryTrueDofs(_boundary_dofs);

  // Build the linear form
  _b = new mfem::ParLinearForm(&(*fes));
  _b->AddDomainIntegrator(new mfem::DomainLFIntegrator(_one));
  _b->Assemble();

  // Build the bilinear form
  _a = new mfem::ParBilinearForm(&(*fes));
  _a->AddDomainIntegrator(new mfem::DiffusionIntegrator);
  _a->Assemble();

  // Form the linear system
  _a->FormLinearSystem(_boundary_dofs, *grid_function, *_b, _problem_operator, _X, _B);
}
```

The solve method solves the linear/non-linear system that has been setup and passes the
data to the mfem `GridFunctions` so that the post-processors can view the results.
During construction as the `Moose::MFEM::ProblemOperator` class was constructed as well,
the class inherits a reference to the `MFEMProblem` namely `_problem_data` it can be used

> to acces `MFEMProblem` members.

Reviewer: ```suggestion
to access `MFEMProblem` members.
```

```cpp
void CustomDummyProblemOperator::Solve() override
{
  // Set the operator and solve the equation
  _problem_data.jacobian_solver->SetOperator(*_problem_operator);
  _problem_data.jacobian_solver->GetSolver().Mult(_B, _X);

  // Set the data in the grid function
  const std::string grid_function_name = "var0";
  auto grid_function = _problem_data.gridfunctions.GetShared(grid_function_name);
  grid_function->SetFromTrueDofs(_X);
};

```

> Once the `ProblemOperator` has been built the `ProblemComposer` class is needed,

Reviewer 2: Not built - written. `ProblemComposer` objects are *built* before `ProblemOperator` objects (since they build them!)

this class is constructed by `MFEMProblem` and called by the executioner. The composer class
must inherit from `ProblemComposerBase` making it an `MFEMObject` and by proxy a 
`MooseObject` thus it has a fixed signature constructor and destructor, it has one method
that has a fixed signature that is called by the executioner. All of the code must
be within the `Moose::MFEM` namespace for simplicity. An example minimal class looks like:

```cpp
namespace Moose::MFEM
{
class CustomDummyProblemComposer : public ProblemComposerBase
{
private:
  type1 param1;
  .
  .
  .

public:
  static InputParameters validParams();

  CustomDummyProblemComposer(const InputParameters & parameters);

  ~CustomDummyProblemComposer() = default;

>   /// Returns a pointer to the operator's equation system.

Reviewer: No.
